test/firewallService.test.ts

```diff
@@ -0,0 +1,83 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import * as vscode from 'vscode';
+import * as TypeMoq from 'typemoq';
+import SqlToolsServiceClient from '../src/languageservice/serviceclient';
+import { FirewallService } from '../src/firewall/firewallService';
+import { HandleFirewallRuleRequest, HandleFirewallRuleResponse,
+    CreateFirewallRuleRequest, CreateFirewallRuleResponse, HandleFirewallRuleParams } from '../src/models/contracts/firewall/firewallRequest';
+import VscodeWrapper from '../src/controllers/vscodeWrapper';
+import { assert } from 'chai';
+import { AzureSession } from '../src/models/interfaces';
+
+
+suite('Firewall Service Tests', () => {
+    let firewallService: FirewallService;
+    let client: TypeMoq.IMock<SqlToolsServiceClient>;
+    let vscodeWrapper: TypeMoq.IMock<VscodeWrapper>;
+
+    setup(() => {
+        client = TypeMoq.Mock.ofType(SqlToolsServiceClient, TypeMoq.MockBehavior.Loose);
+        let mockHandleFirewallResponse: HandleFirewallRuleResponse = {
+            result: true,
+            ipAddress: '128.0.0.0'
+        };
+        let mockCreateFirewallRuleResponse: CreateFirewallRuleResponse = {
+            result: true,
+            errorMessage: ''
+        };
+        client.setup(c => c.sendResourceRequest(HandleFirewallRuleRequest.type, TypeMoq.It.isAny())).returns(() => Promise.resolve(mockHandleFirewallResponse));
+        client.setup(c => c.sendResourceRequest(CreateFirewallRuleRequest.type, TypeMoq.It.isAny())).returns(() => Promise.resolve(mockCreateFirewallRuleResponse));
+        vscodeWrapper = TypeMoq.Mock.ofType(VscodeWrapper, TypeMoq.MockBehavior.Loose);
+        let mockSession: AzureSession = {
+            environment: undefined,
+            userId: 'test',
+            tenantId: 'test',
+            credentials: undefined
+        };
+        let mockExtension: vscode.Extension<any> = {
+            id: '',
+            extensionKind: undefined,
+            extensionPath: '',
+            isActive: true,
+            packageJSON: undefined,
+            activate: undefined,
+            exports: {
+                sessions: [mockSession]
+            }
+        };
+        vscodeWrapper.setup(v => v.azureAccountExtension).returns(() => mockExtension);
+        firewallService = new FirewallService(client.object, vscodeWrapper.object);
+    });
+
+    test('isSignedIn Test', () => {
+        let isSignedIn = firewallService.isSignedIn;
+        assert.isNotTrue(isSignedIn, 'Firewall Service should not be signed in initially');
+        firewallService.isSignedIn = true;
+        assert.isTrue(firewallService.isSignedIn, 'Firewall Service should be signed in once set');
```

Comment:
Not sure what the value in testing a getter/setter like this would be unless it's doing something more complicated in the function. I assume there's something else that sets the isSignedIn property once it detects that the user is signed in. Why aren't you testing that that sets it correctly?

Reply:
I wanetd 100% code coverage for files so I need the getters and setters.

Reply:
Code coverage should include code called further down the stack. So if you make other tests that change these that should get you that coverage as well - you don't have to test every method directly. 

Adding tests for code coverage is fine - but there's so much more useful tests you could be doing here. As an example I see promptForAccountSignIn sets this value to true if the account is signed in. That seems like a more valuable thing to test - that after calling promptForAccountSignIn when vscode reports that the account is signed in that the service is also updated correctly. Since that's a condition which could easily change and is probably important for your app to function. (and also conveniently tests that setting isSignedIn to true will result in that reporting as true correctly). 

Anyways I'm not going to go through and critique every single test - but adding tests just for the sake of pushing a number higher isn't really what we should be aiming for. Every test should have a purpose and the goal here should be to make every test cover a useful piece of functionality that you want to verify works as expected.

Reply:
As you know these tests are unit tests. Some things like signing in is not possible without integration tests. To get full coverage, I had to use the getters/setters. It would be great to do everything in an ideal way, but when that's not possible, you need to resort to this. The goal of this PR is to get code coverage up to 70%. Adding integration tests is not in the scope of this PR and so you'll see getters and setters all around the code which need connections etc. for code coverage like any other unit test has.

Reply:
That's what mocks are for 😄